[iOS] Support visually contiguous bidi selection behavior for more text interactions

whsieh opened the pull request:
#### eb23c6ac8bad7d2cbd9cb6e03a41fabb2cbbea68
<pre>
[iOS] Support visually contiguous bidi selection behavior for more text interactions
<a href="https://bugs.webkit.org/show_bug.cgi?id=284303">https://bugs.webkit.org/show_bug.cgi?id=284303</a>
<a href="https://rdar.apple.com/141163441">rdar://141163441</a>

Reviewed by Aditya Keerthi.

Deploy visually-contiguous bidi text selection support for the following text interactions:

• Tap-and-a-half gesture
• Tap-and-drag gesture
• Trackpad (mouse) drag

While the user is actively modifying text selections, we ensure that the text selection is visually
contiguous across bidi boundaries by having `shouldDrawVisuallyContiguousBidiSelection()` return
`true`. When the text interaction is finished, we then adjust the bidi selection boundaries (if
needed) to be both visually and logically contiguous.

To achieve this, we refactor some of the existing logic to toggle visually contiguous bidi selection
behavior on `WebPage`, such that we keep track of the set of active text interactions using a new
`OptionSet&lt;TextInteractionSource&gt;` on `WebPage`. When the first flag is added, we enter visually-
contiguous bidi selection mode, and when we remove the last flag, we automatically adjust the
selection to nearest visually- and logicaly-contiguous boundaries.

See below for more details.

* LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection-3-expected.txt: Added.
* LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection-3.html: Added.

Add a layout test to exercise the tap-and-drag text interaction. I opted for this one, since
trackpad/mouse simulation isn&apos;t currently supported in WebKitTestRunner, and tap-and-half and tap-
and-drag both exercise the same codepath (but the latter is more simpler (and probably more
reliable) to trigger.

* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/ios/GestureTypes.h:
* Source/WebKit/Shared/ios/GestureTypes.serialization.in:

Add a new enum-class-backed option set type to represent whether the text interaction is driven by
mouse or touch interactions.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didBeginTouchPoint):

Rename `resetPotentialTapSecurityOrigin` to `didBeginTouchPoint`. In addition to resetting the tap
security origin, this now also sets the `m_hasAnyActiveTouchPoints` flag (see below).

(WebKit::WebPageProxy::resetPotentialTapSecurityOrigin): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _touchEventsRecognized]):
(-[WKContentView updateSelectionWithExtentPoint:withBoundary:completionHandler:]):

Pass in whether or not the selection update is triggered by a mouse- or touch-based interaction.
Note that this is an option set (despite the fact that this specific IPC message currently only
sends a single value) because the backing enum class type will be stored as an option set in
`WebPage` anyways.

* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::updateSelectionWithExtentPointAndBoundary):
(WebKit::WebPageProxy::didReleaseAllTouchPoints):

Add a new IPC call to tell the webpage that all touch points have been released.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::mouseEvent):
(WebKit::WebPage::didBeginTouchPoint):

Update `m_hasAnyActiveTouchPoints` in addition to resetting the potential tap security origin.

(WebKit::WebPage::didCommitLoad):

Reset `m_hasAnyActiveTouchPoints` and `m_activeTextInteractionSources` when committing a load.

(WebKit::WebPage::resetPotentialTapSecurityOrigin): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::shouldDrawVisuallyContiguousBidiSelection):

Make this return whether or not `m_activeTextInteractionSources` has any flags set.

(WebKit::WebPage::addTextInteractionSources):
(WebKit::WebPage::removeTextInteractionSources):

Add helpers to add or remove active text interaction sources.

(WebKit::WebPage::updateSelectionWithTouches):
(WebKit::WebPage::updateSelectionWithExtentPointAndBoundary):

Update `m_activeTextInteractionSources` here, by adding the incoming text interaction sources. Note
that this is invoked in response to both tap-and-a-half and tap-and-drag text interaction gestures
as well as mouse drag, but (due to implementation details in UIKit) may be invoked before or after
all touch points have been released in the former 2 cases (see below).

(WebKit::WebPage::didReleaseAllTouchPoints):

This is called when all of the active touch points are released from the web view. Unset
`m_hasAnyActiveTouchPoints` and clear the active touch-based text interaction flag.

(WebKit::WebPage::setShouldDrawVisuallyContiguousBidiSelection): Deleted.

Replaced by `(add|remove)TextInteractionSources`.

Canonical link: <a href="https://commits.webkit.org/287577@main">https://commits.webkit.org/287577@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2563736f9073c34d4fe75c9b674be8f372b16b4a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80110 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59112 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33522 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84627 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31086 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68173 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7410 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62635 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20454 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83179 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/52706 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72964 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42939 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/79567 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50034 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27118 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29546 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71162 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27625 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86059 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7330 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5191 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70906 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7505 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68803 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70142 "Found 2 new API test failures: /TestWebKit:WebKit.DocumentStartUserScriptAlertCrashTest, /TestWebKit:WebKit.InjectedBundleFrameHitTest (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17472 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14141 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13086 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7294 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12811 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7133 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10653 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8938 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3586,6 +3586,11 @@ void WebPage::mouseEvent(FrameIdentifier frameID, const WebMouseEvent& mouseEven
+
+#if PLATFORM(IOS_FAMILY)
+    if (m_page->settings().visuallyContiguousBidiTextSelectionEnabled() && mouseEvent.type() == WebEventType::MouseUp)
+        removeTextInteractionSources(TextInteractionSource::Mouse);
```
Is this intentionally gated on the `visuallyContiguousBidiTextSelectionEnabled` setting? The call to `addTextInteractionSources` isn't, so this probably shouldn't be either?

whsieh replied:
Oh, good catch — I forgot to remove the settings-check here (I'd originally made this option set specific to this feature, but then changed my mind when I realized it could be useful for other things in the future)

whsieh committed:
[iOS] Support visually contiguous bidi selection behavior for more text interactions
https://bugs.webkit.org/show_bug.cgi?id=284303
rdar://141163441

Reviewed by Aditya Keerthi.

Deploy visually-contiguous bidi text selection support for the following text interactions:

• Tap-and-a-half gesture
• Tap-and-drag gesture
• Trackpad (mouse) drag

While the user is actively modifying text selections, we ensure that the text selection is visually
contiguous across bidi boundaries by having `shouldDrawVisuallyContiguousBidiSelection()` return
`true`. When the text interaction is finished, we then adjust the bidi selection boundaries (if
needed) to be both visually and logically contiguous.

To achieve this, we refactor some of the existing logic to toggle visually contiguous bidi selection
behavior on `WebPage`, such that we keep track of the set of active text interactions using a new
`OptionSet<TextInteractionSource>` on `WebPage`. When the first flag is added, we enter visually-
contiguous bidi selection mode, and when we remove the last flag, we automatically adjust the
selection to nearest visually- and logicaly-contiguous boundaries.

See below for more details.

* LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection-3-expected.txt: Added.
* LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection-3.html: Added.

Add a layout test to exercise the tap-and-drag text interaction. I opted for this one, since
trackpad/mouse simulation isn't currently supported in WebKitTestRunner, and tap-and-half and tap-
and-drag both exercise the same codepath (but the latter is more simpler (and probably more
reliable) to trigger.

* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/ios/GestureTypes.h:
* Source/WebKit/Shared/ios/GestureTypes.serialization.in:

Add a new enum-class-backed option set type to represent whether the text interaction is driven by
mouse or touch interactions.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didBeginTouchPoint):

Rename `resetPotentialTapSecurityOrigin` to `didBeginTouchPoint`. In addition to resetting the tap
security origin, this now also sets the `m_hasAnyActiveTouchPoints` flag (see below).

(WebKit::WebPageProxy::resetPotentialTapSecurityOrigin): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _touchEventsRecognized]):
(-[WKContentView updateSelectionWithExtentPoint:withBoundary:completionHandler:]):

Pass in whether or not the selection update is triggered by a mouse- or touch-based interaction.
Note that this is an option set (despite the fact that this specific IPC message currently only
sends a single value) because the backing enum class type will be stored as an option set in
`WebPage` anyways.

* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::updateSelectionWithExtentPointAndBoundary):
(WebKit::WebPageProxy::didReleaseAllTouchPoints):

Add a new IPC call to tell the webpage that all touch points have been released.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::mouseEvent):
(WebKit::WebPage::didBeginTouchPoint):

Update `m_hasAnyActiveTouchPoints` in addition to resetting the potential tap security origin.

(WebKit::WebPage::didCommitLoad):

Reset `m_hasAnyActiveTouchPoints` and `m_activeTextInteractionSources` when committing a load.

(WebKit::WebPage::resetPotentialTapSecurityOrigin): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::shouldDrawVisuallyContiguousBidiSelection):

Make this return whether or not `m_activeTextInteractionSources` has any flags set.

(WebKit::WebPage::addTextInteractionSources):
(WebKit::WebPage::removeTextInteractionSources):

Add helpers to add or remove active text interaction sources.

(WebKit::WebPage::updateSelectionWithTouches):
(WebKit::WebPage::updateSelectionWithExtentPointAndBoundary):

Update `m_activeTextInteractionSources` here, by adding the incoming text interaction sources. Note
that this is invoked in response to both tap-and-a-half and tap-and-drag text interaction gestures
as well as mouse drag, but (due to implementation details in UIKit) may be invoked before or after
all touch points have been released in the former 2 cases (see below).

(WebKit::WebPage::didReleaseAllTouchPoints):

This is called when all of the active touch points are released from the web view. Unset
`m_hasAnyActiveTouchPoints` and clear the active touch-based text interaction flag.

(WebKit::WebPage::setShouldDrawVisuallyContiguousBidiSelection): Deleted.

Replaced by `(add|remove)TextInteractionSources`.

Canonical link: https://commits.webkit.org/287577@main
